llvm/lib/Transforms/IPO/SampleProfileMatcher.cpp

```diff
@@ -863,6 +883,39 @@ bool SampleProfileMatcher::functionMatchesProfile(Function &IRFunc,
   return Matched;
 }
 
+void SampleProfileMatcher::UpdateSampleLoaderWithRecoveredProfiles() {
+  DenseSet<StringRef> RecoveredFuncs;
+  // Update FuncNameToProfNameMap and SymbolMap.
+  for (auto &I : FuncToProfileNameMap) {
+    assert(I.first && "New function is null");
+    FunctionId FuncName(I.first->getName());
+    RecoveredFuncs.insert(I.second.stringRef());
+    FuncNameToProfNameMap->emplace(FuncName, I.second);
+
+    // We need to remove the old entry to avoid duplicating the function
+    // processing.
+    SymbolMap->erase(FuncName);
+    SymbolMap->emplace(I.second, I.first);
+  }
+
+  // Read the top-level profiles for the recovered function profiles. This is
+  // because in extended binary format it only loads the top-level profile for
+  // the functions in the new build but not the recovered functions which is
+  // from the old build.
+  SampleProfileMap TopLevelRecoveredProfiles;
+  Reader.readOnDemand(RecoveredFuncs, TopLevelRecoveredProfiles);
+  auto &Profiles = Reader.getProfiles();
+  for (auto &I : TopLevelRecoveredProfiles) {
+    LLVM_DEBUG(dbgs() << "Top-level function " << I.second.getFunction()
+                      << " is recovered and re-read by the sample reader.\n");
+    auto &Ctx = I.second.getContext();
+    assert(Profiles.find(Ctx) == Profiles.end() &&
+           "Top level profile is found for the unused profile");
+    Profiles.create(Ctx) = std::move(I.second);
```

Comment:
How about we still populate the newly loaded profiles into the same `SampleProfileReader::Profiles` map? This would make reader APIs simpler and more consistent, and we won't need the assertion above either.

Reply:
What I was suggesting is that the `read` no longer need a `SampleProfileMap` parameter to hold newly loaded profile. I'm still seeing `read` having that parameter in the latest version?